Transactional selector cache

jrajahalme opened the pull request:
Make SelectorCache identity updates transactional. This allows changes to selections of different selectors to appear at once on a new version published by the selector cache after all the changes have been completed, which allows removal of the CIDR index within MapState, making computation of CIDR heavy policies ~45% faster, mainly due to reduction of allocations by ~37%.

This change allows SelectorCache users to rely on all the selectors to be in sync with each other, so that if a given identity is selected by multiple selectors, that identity appears in the result of the `GetSelections()` call for all of them. Previously it was possible that at the time of the respective GetSelections() calls, some of them would have failed to include the new identity, due to the update of that selector still being in progress.

With this change the visibility of new identities in the selections of any cached selector is delayed until all affected selectors have been updated.

Incremental updates are synchronized after each round of identity modifications in the selector cache so that the updates on all affected MapState entries appear during the same round of updates, essentially making incremental updates transactional as well. The consumer of incremental updates gets the version of the selector cache as of after the changes, so that a consistent NetworkPolicy can be computed for Envoy as well.

Now that SelectorCache GetSelections() and incremental updates are both transactional, we can rely on a covering CIDR selector to also select any covered new identities. For example, given a L3-only WORLD deny rule and an toFQDNs allow rule, and newly allocated FQDN identities are also selected by the deny WORLD selector, and we can rely on these new
identities appearing on those selectors "at the same time", so that on any given round of DistillPolicy or ConsumeMapChanges, any new identities appear on all affected selectors, and conversely, and removed identities are likewise removed from all affected selectors.

Given the above, we do not need to go look for non-wildcard superset/subset identities, as keys with them will be added (or have already been added) or removed on the same round and the effect of them will be present in the computed MapState. In the above example, even if on a given round we first add an allow rules the new FQDN-derived identity, we will later replace it with a deny entry based on the same ID being selected by the (deny) WORLD selector. Conversely, if on a given round we add the deny entry first, the allow entry will be naturally bailed due to the existing deny entry.

Following PR(s) have been carved out to merge separately:
- #34343
- #34346
- #34347
- #34349
- #34562
- #34679

jrajahalme committed:
policy: Test CIDRs with all covered identities

Real CIDR (and WORLD) selectors select all covered identities, so that
when we generate MapState for a CIDR or a WORLD rule, entries for all the
covered identities are created. Mimic this in unit tests as well.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Avoid adding unnecessary derivative deny entries

Deny entries with broader (i.e., less specific) port-protocols interact
with superset allow entries only if the allow entry is an L3-wildcard. In
the case of the L3-wildcard allow, the deny entry is by definition a
non-L3-wildcard (its identity is a subset of the allow identity), and
datapath does not relate two non-wildcard L3-entries in any way. Given
this we need to add additional L3L4-deny entries only if the allow entry
in this scenario has an L3-wildcard key.

Tests are adjusted to not expect the unnecessary derivative deny entries.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Rename test variables

Rename test variables in a separate commit first to highlight functional
changes later.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Return ChangeState from ConsumeMapChanges

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+// GetHandle returns a Handle for the current version and a closer for releasing old versions held
+// for this version.
+// 'name' is for debugging purposes only
+func (v *Manager) GetHandle(name string) (handle Handle, closer Closer) {
```
Would it not be more idiomatic in Go to have `handle` be a standard `io.Closer` (https://pkg.go.dev/io#Closer) and hence the code of the closer func to be a method on handle?

jrajahalme replied:
Thanks for the review, I've changed this since your review. I thought about conforming to `io.Closer`, but hesitated since this is no "IO"' as such. The method is now called `Release()`, but given your suggestion I'll rename it as `Close` instead.

jrajahalme replied:
done!

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "versioned")
```
let's not add more package global loggers, please. Make it a field on the manager?

dylandreimerink replied:
To add to that, could we use slog for new code instead of logrus, given we are transitioning

jrajahalme replied:
Are there any good examples around?

joamaki replied:
grep for `slog.Logger`. (almost) everything at least in pkg/datapath/linux uses slog now. tl;dr: depend on `*slog.Logger` (e.g. a constructor (Provide) can grab this, then pass it as parameter to NewVersionManager). There's no formatted methods, instead: ```log.Info("Foobar", "id", id, logfields.Error, err)``` etc.
The nice thing doing it this way is that then the logging happens in the context of the module using this, so its easier to figure out what it's related to (not that there's I guess many use-cases for this...)

joamaki replied:
https://go.dev/blog/slog

jrajahalme replied:
None of the logs come via the manager... ended up using the `slog` without defining a new global. Not sure how logging error levels are passed to it though.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+	return newHandle(maxVersion)
+}
+
+func (h Handle) getVersion() Version {
```
just `version` is more idiomatic Go.

jrajahalme replied:
Thanks, will change.

jrajahalme replied:
done :-)

bimmlerd commented on pkg/container/versioned/value_test.go:
```diff
@@ -0,0 +1,289 @@
+	handle, closer := cv.GetHandle("empty")
+	assert.Empty(t, GetValue(&value1, handle))
+	assert.True(t, closer())
+	// 2ns call does nothing
```
nit: typo; `2nd`

jrajahalme replied:
thanks, will fix.

jrajahalme replied:
fixed, thanks!

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+	return current, current + 1
+}
+
+func (v *Manager) PublishVersion(current, new Version) bool {
```
It wasn't immediately clear to me why publish would need the current version. Please add a doc comment, and consider using a name which evokes CompareAndSwap a bit more?

jrajahalme replied:
Yeah, if there are no user bugs the current version is not needed, but it is taken to enforce "no concurrent" updates from multiple writers. Will add a comment.

jrajahalme replied:
Simplified to take a new `versioned.NextVersion` type defined just for this purpose. With another `KeepVersion` type I've made the main version type private (`version`).

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+// when an overlapping entry is added, or a value is removed.
+//
+// 'version' must be later than any the current version visible to the readers.
+// Must be called from a single writer!
```
> Must be called from a single writer!

What does this mean exactly? Does this mean that it _must_ always be the same thread of execution which calls `SetValueAtVersion` or that writers must be serialized?

jrajahalme replied:
The latter, will clarify.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+// All values are added with "infinite" validity, which is then truncated
+// when an overlapping entry is added, or a value is removed.
+//
+// 'version' must be later than any the current version visible to the readers.
```
What happens otherwise? Should this return an error? Or panic?

jrajahalme replied:
IMO panic would be the right thing to do, but we have a strict "no panics" policy. One way to resolve this would be to bring the given version forward and insert with a warning log?

jrajahalme replied:
added an error return.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+	for node != nil {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			log.Errorf("AddValueAtVersion may not add values at versions lower than those already added! (%d < %d). Stacktrace: %s", version, node.versions.first, stacktrace)
```
I don't think "just" logging is a resilient way of dealing with this, why not return an error?

jrajahalme replied:
Ok, will add error return and check on the call sites :-)

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+
+// Value is a container for versioned value, implemented as a lock-free linked list
+type Value[T any] struct {
+	// versionedNodes are non-overlapping and in sorted order
```
nit: `versionedNodes` no longer exists, probably `head`?
sorted by version I assume, but ascending or descending?

jrajahalme replied:
ascending by version, and `valueNode`s (after a recent rename `node` -> `valueNode`.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+// Must be called from a single writer!
+func RemoveValueAtVersion[T any](v *Value[T], version Version) {
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if version < node.versions.first {
```
Same as before, not sure "just" logging is enough here. Return an error?

jrajahalme replied:
will add error return. Just logging has been pretty effective during development, though.

jrajahalme replied:
done.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+
+// TrimFrontPairSlice trims the first 'n' Pairs off of the PairSlice 's', returning 's'.
+// If 's' gets empty it is reallocated if its capacity exceeds 'capacity'.
+func TrimFrontPairSlice[T any](s PairSlice[T], n int, capacity int) PairSlice[T] {
```
The Go runtime is pretty good at managing slice memory, how come we need to manage this ourselves here?

jrajahalme replied:
TIL, I thought large capacity would stick forever! Do you have a reference for this?

bimmlerd replied:
Sorry, to be clear - Go cannot release the underlying array while there is a slice referencing it. In that sense, a large cap will stick around, even if `len` is short.

I don't yet understand the usage pattern for `PairSlices` well enough to get why this would be benefitial compared to just `s[n:]`; though. Or maybe a
```go
if len(s) == n {
    return nil
}
return s[n:]
```
 If we're continuously writing towards at the tail and releasing at the front, should we be using a ring buffer?

jrajahalme replied:
Do we have a nice "slice like ring buffer", It would need to expand when needed, as hitting a capacity limit is not acceptable.

Maybe we should just always copy the content to the beginning of the slice?

Or maybe just holding on to the storage should be fine? It should get empty very often, and then the runtime should start filling it from the beginning of the storage, correct?

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+			return false
+		}
+
+		v.mutex.Lock()
```
`defer v.mutex.Unlock()`?

jrajahalme replied:
yes

jrajahalme replied:
no for the current version, as we want to unlock before the end of the function.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+
+// AppendPair appends a pair of 'version' and 'value' to PairSlice 's', returning 's'
+func AppendPair[T any](s PairSlice[T], version Version, value T) PairSlice[T] {
+	s = append(s, Pair[T]{
```
just `return append(s, ...)`
and since then it's a single line function, is this really worth having?

jrajahalme replied:
lets see...

jrajahalme replied:
Removed unnecessary types related to this, but a function for this is needed so that the members can be private. Not sure it that is important, though.

bimmlerd commented on pkg/container/versioned/value_test.go:
```diff
@@ -0,0 +1,289 @@
+	// new handle sees the new value
+	handle3, closer3 := cv.GetHandle("third")
+	v = GetValue(&value1, handle3)
+	assert.Len(t, v, 3)
```
The length assertions are redundant if equality is checked afterwards, and the error produced is probably better in the equal check?

jrajahalme replied:
yes.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+
+// GetValue returns value of type 'T' valid for the given handle, or an empty value if none is
+// found.
+func GetValue[T any](v *Value[T], handle Handle) T {
```
This API reads a bit strange to me, possibly because it's ported from C? In Go, I'd expect something like

`val := value.At(version)` instead of `val := GetValue(value, version)`

To me, just `Handle` is also not descriptive enough, maybe `VersionRef` or `VersionHandle` would communicate more clearly that this is a reference to a version. Alternatively, we could make a current `Version` be `VersionNum` or a `Revision` and then have the `Handle` be the `Version`, but I'm not sure that's clearer.

jrajahalme replied:
Generics can not be receivers, AFAIK. We could create a shims for specific types if need be, though.

Renamed as `ValueAtVersion()`.

jrajahalme replied:
Renamed `Handle` as `VersionHold` and made it to accessed via a pointer. The value of `VersionHold` is not to be copied around, as that would mess with tracking of stale values. Also added a finalizer to catch un-closed `VersionHolds()`.

joamaki replied:
I'm a bit confused why we can't do `value.At`. We are not introducing new type parameters, so this type should be fine:
```
func (v *Value[T]) At(h VersionHold) T { ... }
```

joamaki replied:
Actually all the functions that take Value[T] should just be normal methods on it.

jrajahalme replied:
Yes, thanks for pointing this out. I'm learning Go generics "on-the-go", and faced a compiler error relating to something like this. It must have been that the method in question introduced a new generic type, and I thought the error to meant that no generic type can be a receiver...

bimmlerd replied:
Yeah, here's the reasoning - essentially you can't _add_ more type parameters in a method, but the method can use the type parameters of its receiver.
https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#no-parameterized-methods

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+	past  atomicVersion // first version this value is invalid for
+}
+
+func (r *versionRange) Contains(version Version) bool {
```
`contains`; no need to export it

jrajahalme replied:
unexported

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+}
+
+// Manager manages the versioning for multiple VersionedValues.
+// It is to be stored wherever VersionedValues are stored
```
What does this mean?

jrajahalme replied:
Should read `versioned.Value`s. But generally `Manager` is used to share a single version number space across multiple `versioned.Value`s.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,337 @@
+			keepVersion = v.handles[0]
+		}
+		// 'keepVersion' is the current version if there are no outstanding handles
+		if keepVersion == invalidVersion {
```
isn't this just `else`? Or is having `v.handles[0]` be invalidVersion expected?

jrajahalme replied:
not really expected, will simplify.

bimmlerd reviewed (commented):
mostly looked at the `versioned` package, I think the API could be a bit clearer

joamaki commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,375 @@
+	cleanedVersion Version
+}
+
+func NewVersionManager(cleaner func(keepVersion Version)) *Manager {
```
I wonder if something like `VersionSet` would be a better name here?

jrajahalme replied:
I also despise the generic `Manager` as a name, but it is not a container itself, so `Set` would be wrong IMO.

joamaki replied:
Right yeah, `Set` seems wrong indeed. Maybe `Coordinator` could work too, but that's as generic as manager.

jrajahalme replied:
I just thought that `Coordinator` would be better, came here and saw you suggested it. Reading that days ago must have primed me to this :-)

joamaki commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,375 @@
+}
+
+// Pair is a struct of a version and any type T
+type Pair[T any] struct {
```
`Versioned[T]` might be more descriptive?

jrajahalme replied:
I like this, Done!

joamaki commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,425 @@
+	count   int
+}
+
+// Manager defines a commont version number space for multiple versioned.Values,
```
typo in `commont`

jrajahalme replied:
fixed, thanks!

joamaki commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,425 @@
+	cleanedVersion version
+}
+
+// NewVersionManager returns a new VersionManager
```
Match this with `Manager`? e.g. `NewManager`. Using `New` would be more go-like, but that's probably confusing.

jrajahalme replied:
Changed to `NewManager`. When there are multiple types in a package a bare `New` would indeed be confusing.

squeed commented on pkg/endpoint/policy.go:
```diff
@@ -141,6 +141,12 @@ func (e *Endpoint) updateNetworkPolicy(proxyWaitGroup *completion.WaitGroup) (re
 	}
 
+	// Need a valid handle to be able to update the network policy, get one if needed
+	if !e.desiredPolicy.VersionHold.IsValid() {
```
When would this not be valid? Couldn't this cause us to miss incremental updates?

jrajahalme replied:
The handle taken for the full policy computation is closed as soon as possible after that process finishes to release resources held for that specific version. Incremental updates are not tied into that version handle itself, but we have tight synchronization at the creation of the EndpointPolicy that gives as all incremental updates after the version for which the EndpointPolicy was originally created from.

jrajahalme replied:
So when we receive incremental updates and this function is called outside of the original full policy regeneration path, this handle would be expected to be invalid. Either way, we have a newer handle we receive from `ConsumeMapChanges()` that we must use while updating the network policy that we swap in.

This specific change is updated in the last commit, moving this logic to happen right after `ConsumeMapChanges` call like this:
```
	version, changes := e.desiredPolicy.ConsumeMapChanges()

	// version is not valid if there are no changes
	if version.IsValid() {
		// Swap the new version in place of a valid version handle, or close when done
		if e.desiredPolicy.VersionHandle.IsValid() {
			e.desiredPolicy.VersionHandle.Close()
		} else {
			defer e.desiredPolicy.Ready()
		}
		e.desiredPolicy.VersionHandle = version
	}
```

Here `ConsumeMapChanges` has just applied the incremental policy changes to the endpoint's desired policy, bringing it up to the new `version` received from `ConsumeMapChanges`. If there were any changes, we replace the version handle at `e.desiredPolicy` with the new one, and release it via `e.desiredPolicy.Ready()` unless `.desiredPolicy.VersionHandle` was already valid (so we are in the middle of a full policy compute), in which case it is left "open" for any use after this call.

Looking at this now, I feel like the `ConsumeMapChanges`, which is a method of the `EndpointPolicy`, itself should take care of the update of a valid version handle, that might make this a bit clearer.

Overalll, when applying incremental policy updates, we want to update `e.desiredPolicy.VersionHandle` and release it if it was not open/valid when we started.

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -105,15 +108,24 @@ type SelectorCache struct {
 }
 
+// GetCurrentVersionHold returns a VersoionHold for the current version.
+// The returned VersionHold must be closed with Close()
```
I wonder if we can add a lint rule that detects un-closed holds? I'm not familiar with this sort of analysis.

jrajahalme replied:
I added Endpoint.buildLock/Unlock functions that are used instead of calling `e.buildMutex.Lock` or `e.buildMutex.Unlock` directly, and added "linting" there. IMO the build mutex unlock is the perfect spot for this.

squeed commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+// Close releases the held version for removal once no hold for it are no longer held.
+// This may not be called while holding any locks that the 'closer' function passed to the
+// coordinator may take!
+func (h *VersionHold) Close() (err error) {
```
Idea: in debug mode, add a finalizer that panics if Close() was not called before the value was GCd.

jrajahalme replied:
I like this, but do we want the panic to happen if a customer runs Cilium in debug mode? The warning log is there to draw attention to this, and the stacktrace provides the same information a panic would.

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -202,19 +205,7 @@ func NewSelectorCache(ids identity.IdentityMap) *SelectorCache {
-			idSel.selections.RemoveBefore(keepVersion)
-		}
-	})
+	sc.versioned = versioned.NewCoordinator(sc.oldVersionCleaner)
```
minor nit: this changes twice in the PR; maybe we can simplify slightly.

jrajahalme replied:
I'll try to reorganize to limit multiple changes to the same code blocks within the PR.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+type NextVersion version
+
+var (
+	errInvalidVersion  = errors.New("invalid version")
```
If these are not exported, other packages cannot write `errors.Is(err, errInvalidVersion)` - is that an intentional choice or should these errors be public API?

jrajahalme replied:
This was an oversight, thanks for noticing, exporting this.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+//
+// A special form with a nil coordinator, which is returned by Latest(), always finds the latest
+// versions and does not keep any versions from being released.
+type VersionHold struct {
```
Should this struct be exported? It doesn't seem like it's possible to construct one from outside the package, since all fields are unexported.

jrajahalme replied:
So refer to them via an interface instead?

jrajahalme replied:
So basically other packages need to be able to hold references (pointers) to `VersionHandle`, so some type needs to be exported for that purpose, right? Adding an interface would add functionality that is not needed, so exporting this type, even when it has no exported fields should be fine?

bimmlerd replied:
no, you can have an unexported struct with exported methods. https://go.dev/play/p/lFYPMidT6qW

jrajahalme replied:
Moreover, `VersionHandle` is a struct, and it is deliberately made to have no exported fields. Since it is accessed via a pointer, an exported interface could be used instead. But if it were accessed by value, then that would not be an option.

bimmlerd replied:
But the proverb goes `Accept interfaces return structs` 🙃 The point is that you're supposed to define interfaces at the site of the consumer, not the producer.

bimmlerd replied:
Ah - there is one reason which makes it slightly more convenient to have it exported I suppose:

If unexported, and you want to declare functions taking `*versionHold` _outside_ of the `versioned` package you'd have to declare an interface _in that package_ which `*versionHold` implicitly implements, whereas with an exported struct you can just use that. Mostly a preference thing, I suppose - the former has the benefit of the producer being able to swap out the implementation transparently if it conforms to the interface still. 

```go
type version interface {
    Close()
}

func foo(ver version) {} // will happily take *versionHold
```

jrajahalme replied:
Done. Going through the interface required adding some `nil` checks as methods can not be called via a nil interface as they can via nil pointers, where the method then checks for the nil pointer.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+//
+// A special form with a nil coordinator, which is returned by Latest(), always finds the latest
+// versions and does not keep any versions from being released.
+type VersionHold struct {
```
`VersionHold` doesn't read well to me and doesn't invoke the connotation of "keeping something alive". Maybe `VersionReference` (or `VersionRef` if that's too long)? or `VersionHandle`?

jrajahalme replied:
Went with `VersionHandle`.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+
+// versionHoldFinalizer is used to warn about missing Close() calls.
+func versionHoldFinalizer(h *VersionHold) {
+	slog.Error("Hold for SelectorCache version not closed.", logfields.Version, h.version, logfields.LogSubsys, "containers/versioned")
```
This uses the default, unconfigured slog logger. That means that if someone configures the agent to, say, log as JSON, these log lines would show up as just text, since cilium configures an instance of an *slog.Logger. My idea was that if the `VersionHold`s have a reference to the coordinator, they could use that to get access to a logger.
However, it seems that not all holds have a coordinator, so either they all need to hold a pointer to a logger themselves, or not log.

jrajahalme replied:
I moved the logging to the caller, except for the finalizer for which a `Coordinator` reference exists when it needs to log.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+
+// versionHoldFinalizer is used to warn about missing Close() calls.
+func versionHoldFinalizer(h *VersionHold) {
+	slog.Error("Hold for SelectorCache version not closed.", logfields.Version, h.version, logfields.LogSubsys, "containers/versioned")
```
I think this should panic, not error log. We have precendence of catching such programming errors via panic in the resource code, so this should be acceptable.

jrajahalme replied:
I recall being told `panic` is not acceptable. Here we are recovering without crashing the agent and emitting a warning log, so we'll know if this ever happens in the CI, which is as good as panicing.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+}
+
+// NewCoordinator returns a new Coordinator
+func NewCoordinator(cleaner func(KeepVersion)) *Coordinator {
```
Do we really need this constructor or could `Coordinator` be just a struct that has a public `Cleaner` field? The other fields seem like they have the correct zero-value behaviour.

jrajahalme replied:
Removed this, thanks for the suggestion. Test code became simpler since there we can skip setting up the logger.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+
+// GetNextVersion returns NextVersion to be used when adding or removing values.
+//
+// Callers need to coordinate so that a single goroutine is performing modifications at any one
```
Looking at this, I can see the transactionality clearly. Should the transaction be made explicit in the API? Explicit transactions in Go could look like in SQL https://pkg.go.dev/database/sql#Tx or in statedb: https://pkg.go.dev/github.com/cilium/statedb#ReadTxn

jrajahalme replied:
Updated, now it looks like this:
```
// - tx := coordinator.PrepareNextVersion()
//   - value.SetAt(... , tx)
//   - value.RemoveAt(..., tx)
//   - ...
//
// - tx.Commit()
```

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+// value associated with versions earlier than 'keepVersion'.
+// The slice is assumed to be sorted by version in ascending order.
+// Returns the number of calls made.
+func (s VersionedSlice[T]) ForEachBefore(keepVersion KeepVersion, f func(T)) (n int) {
```
With the new range func in Go 1.23, this should be an iterator instead, so that folks can do

```go
for t := range vslice.ForEachBefore(ver) {
   // whatever they want with t
}
```

or even

```go
for v, t := range vslice.ForEachBefore(ver) {
   // whatever they want with t and v
}
```

jrajahalme replied:
Nice! Will do when we get to Go 1.23!

gandro replied:
We now do have v1.23 in tree. Would it make sense to adapt this now or are you planning to do it in a followup?

jrajahalme replied:
Done!

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+			return node.value
+		}
+	}
+	var empty T
```
nit: slightly more elegant to name the return value `empty` so that you don't have to declare it 

```go
func (v *Value[T]) At(hold *VersionHold) (empty T) {
...
return
}
```

jrajahalme replied:
Changed, but with `return empty` which IMO read better :-)

bimmlerd reviewed (changes requested):
One more pass at the version.Value API

biggest blocker for me is the use of the global slog. Also, I think the transaction could be explicit, but maybe passing it through in places makes it inconvenient.

bimmlerd commented on daemon/cmd/endpoint.go:
```diff
@@ -1158,7 +1158,12 @@ func (d *Daemon) GetDNSRules(epID uint16) restore.DNSRules {
 	}
 
-	rules, err := proxy.DefaultDNSProxy.GetRules(epID)
+	// We get the latest consitent view on the DNS rules by getting handle to the latest
```
```suggestion
	// We get the latest consistent view on the DNS rules by getting a handle to the latest
```

jrajahalme replied:
fixed, thanks!

bimmlerd commented on pkg/policy/selectorcache.go:
```diff
@@ -262,6 +290,7 @@ func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, lbls labels.L
 }
 
+// must hold lock for writing
 func (sc *SelectorCache) addSelector(user CachedSelectionUser, lbls labels.LabelArray, key string, source selectorSource) (CachedSelector, bool) {
```
nit: rename to `addSelectorLocked` if the lock must be held?

jrajahalme replied:
Good call, while this aspect was not changed in this PR, we already have `removeSelectorLocked`, so I renamed this as suggested.

bimmlerd commented on pkg/policy/resolve.go:
```diff
@@ -41,6 +43,11 @@ type EndpointPolicy struct {
 
+	// VersionHold represents the version of the SelectorCache 'policyMapState' was generated from.
+	// Changes after this version appear in 'policyMapChanges'.
+	// This is updated when increamental changes are applied.
```
nit `incremental`

jrajahalme replied:
fixed, thanks!

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -834,7 +835,8 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPortProto restore.PortPro
 	for selector, regex := range epAllow {
 		// The port was matched in getPortRulesForID, above.
-		if regex != nil && selector.Selects(destID) && (regex.String() == matchpattern.MatchAllAnchoredPattern || regex.MatchString(name)) {
+		// versioned.AllHandle() always gets the latest version of selectors
```
nit: `AllHandle` -> `Latest` and then maybe the comment is no longer necessary

jrajahalme replied:
Removed, thanks!

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,440 @@
+	version atomicVersion
+
+	// cleaner is called with the earliest version that must be kept
+	cleaner func(KeepVersion)
```
add a note that the cleaner may be called concurrently, which the func needs to handle itself

jrajahalme replied:
Added a comment.

bimmlerd commented on pkg/endpoint/bpf.go:
```diff
@@ -1248,7 +1248,18 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+		return nil
+	}
+	// update a valid handle
+	if e.desiredPolicy.VersionHold.IsValid() {
```
doesn't `Close` already check `IsValid`?

jrajahalme replied:
Yes, but here we need to know the state of the handle before the close. (this code has now moved to `pkg/policy/resolve.go`)

bimmlerd commented on pkg/policy/l4.go:
```diff
@@ -1663,6 +1676,24 @@ func (l4Policy *L4Policy) AccumulateMapChanges(l4 *L4Filter, cs CachedSelector,
+			continue
+		}
+		if option.Config.Debug {
+			log.Debug("SyncMapChanges")
```
Is this a useful debug log?

jrajahalme replied:
It was when I was debugging this, no longer needed, so removed :-)

bimmlerd commented on pkg/policy/l4.go:
```diff
@@ -1598,16 +1616,11 @@ func (l4Policy *L4Policy) AccumulateMapChanges(l4 *L4Filter, cs CachedSelector,
-	// Must take a copy of 'users' as GetNamedPort() will lock the Endpoint below and
-	// the Endpoint lock may not be taken while 'l4.mutex' is held.
+	// Can hold rlock here as neither GetNamedPort() nor LookupRedirectPort() no longer
+	// takes the Endpoint below.
```
nit: `Endpoint lock`

jrajahalme replied:
fixed, thanks!

bimmlerd commented on pkg/policy/selectorcache.go:
```diff
@@ -106,11 +107,25 @@ type SelectorCache struct {
 	userMutex lock.Mutex
 	// userNotes holds a FIFO list of user notifications to be made
 	userNotes []userNotification
+	// notifiedUsers is a set of all notified userNotes
```
it's the set of all notified users, right?

jrajahalme replied:
correct! fixed

bimmlerd commented on pkg/policy/mapstate.go:
```diff
@@ -1970,8 +1971,10 @@ func (ms *mapState) getIdentities(log *logrus.Logger, denied bool) (ingIdentitie
-	mutex   lock.Mutex
-	changes []MapChange
+	mutex       lock.Mutex
+	changez     []MapChange
```
changez?

jrajahalme replied:
Heh :-) Renamed "temporarily" to make sure there were no unwanted referfences to it :-)

jrajahalme replied:
fixed :-)

bimmlerd commented on pkg/policy/resolve_deny_test.go:
```diff
@@ -477,7 +477,7 @@ func TestMapStateWithIngressDenyWildcard(t *testing.T) {
 	defer td.sc.UpdateIdentities(nil, added1, wg)
 	wg.Wait()
-	require.Equal(t, 0, len(policy.policyMapChanges.changes))
+	require.Equal(t, 0, len(policy.policyMapChanges.synced)) // XXX what is this?
```
comment needs cleanup?

jrajahalme replied:
done!

bimmlerd reviewed (commented):
couple more nits

jrajahalme committed:
policy: Rely on CIDR & WORLD selectors selecting all covered identities

Remove the CIDR index from MapState.

Now that SelectorCache GetSelections() and incremental updates are both
transactional, we can rely on a covering CIDR identity also select any
covered new identities. For example, given a L3-only WORLD deny rule and
an toFQDNs allow rule, and newly allocated FQDN identities are also
selected by the deny WORLD selector, and we can rely on these new
identities appearing on those selectors "at the same time", so that on
any given round of DistillPolicy or ConsumeMapChanges, any new identities
appear on all affected selectors, and conversely, and revoved identities
are likewise removed from all affected selectors.

Given the above, we do not need to go look for non-wildcard
superset/subset identities, as keys with them will be added (or have
already been added) or removed on the same round and the effect of them
will be present in the computed mapstate. In the above example, even if
on a given round we first add an allow rules the new FQDN-derived
identity, we will later replace it with a deny entry based on the same ID
being selected by the (deny) WORLD selector. Conversely, if on a given
round we add the deny entry first, the allow entry will be naturally
bailed due to the existing deny entry.

This allows to get rid of the Identities interface as well as that was
used to find a prefix for an ID in the selector cache. We also remove the
validator interface, as maintaining it would be a burden at this point.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Sync incremental updates

Make incremental policy updates available for the endpoint only after all
affected selectors have been updated.

We do this by adding IdentitySelectionCommit() call to the
CachedSelectionUser interface. This is called after calls to
IdentitySelectionUpdated with all the affected selectors have been
completed.

IdentitySelectionCommit() gets a *versioned.Tx argument that is used when
the incremental changes result in actual MapState changes to get the
handle with which the Envoy policy update can be computed.

We also synchronize the insertion of a new EndpointPolicy to a
SelectorPolicy with the selector cache identity updates. This guarantees
that all updates after the initial state are received.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

dylandreimerink commented on pkg/policy/l4.go:
```diff
@@ -1657,6 +1666,17 @@ func (l4Policy *L4Policy) AccumulateMapChanges(l4 *L4Filter, cs CachedSelector,
 		}
 		epPolicy.policyMapChanges.AccumulateMapChanges(cs, adds, deletes, keysToAdd, value)
 	}
+	l4Policy.mutex.RUnlock()
```
nit: Quite a bit of space between the RLock and RUnlock, plus no logic after the unlocking, good candidate to use `defer` to avoid future modification to this function from leaking the lock

jrajahalme replied:
Changed to use `defer` :-)

dylandreimerink commented on pkg/policy/mapstate.go:
```diff
@@ -2013,29 +2017,60 @@ func (mc *MapChanges) AccumulateMapChanges(cs CachedSelector, adds, deletes []id
+		}
+	}
+	mc.changes = nil
+	mc.mutex.Unlock()
```
nit: Same here, I would recommend using a `defer`

jrajahalme replied:
Changed to use `defer` :-)

dylandreimerink commented on pkg/policy/mapstate.go:
```diff
@@ -2049,15 +2084,21 @@ func (mc *MapChanges) consumeMapChanges(policyOwner PolicyOwner, policyMapState
+	mc.version = nil
+
+	mc.synced = nil
 	mc.mutex.Unlock()
```
nit: Same here, large function, easy to add an early return and leaking the lock by accident

jrajahalme replied:
Changed to use `defer` :-)

dylandreimerink reviewed (approved):
LGTM, just some small nits

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,510 @@
+}
+
+// LatestTx refers to maxVersion without having a coordinator, should only be used for testing.
+var LatestTx = &Tx{nextVersion: maxVersion}
```
nit: if you move this to the _test.go file, it will only be possible to use it in tests.

jrajahalme replied:
But only from the `versioned` package itself. This is used also from `pkg/policy`.

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,510 @@
+	return tx.coordinator.getVersionHandle(tx.nextVersion)
+}
+
+func versionHandleCmp(a versionCount, b version) int {
```
nit: I think this can be replaced with `cmp.Compare`

jrajahalme replied:
wouldn't this require 'a' and 'b' to be of the same type, and `comparable`? Also, we do not want to take the `count` in `versionCount` into the comparison.

bimmlerd replied:
err, must've not had my coffee when I wrote that >.< yep!

bimmlerd commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,510 @@
+
+// commit makes a new version of values available to readers
+// and cleans up any possible stale versions
+func (v *Coordinator) commit(version version) error {
```
nit: leftover receiver name `v`, should probably be `c`

bimmlerd commented on pkg/container/versioned/value_test.go:
```diff
@@ -0,0 +1,388 @@
+func (c *testCleaner) waitUntilCleaned(version version) {
+	c.cleanerMutex.Lock()
+	for oldest := c.oldestVersion.load(); version >= oldest; oldest = c.oldestVersion.load() {
+		fmt.Printf("Waiting due to oldest %d < %d version\n", oldest, version)
```
nit: `<=`?

jrajahalme replied:
This would be easier to read if the comparison order was flipped, i.e.:
```
  for ... ; oldest <= version; ... {
```
This would be read like "wait as long as the oldest (non-cleaned) version is less or equal than `version`", or "wait until oldest version is greater than `version`"; after this `version` is no longer available, it is "cleaned off".

bimmlerd replied:
oh I just meant that the printed message is different from the actual check

bimmlerd commented on pkg/container/versioned/value_test.go:
```diff
@@ -0,0 +1,388 @@
+			go func() {
+				version := cv.GetVersionHandle()
+				time.Sleep(time.Duration(rand.IntN(100)) * time.Millisecond)
+				fmt.Printf("Closing version %s\n", version)
```
nit: if you use `t.Logf` in the tests, the logging only shows up when the test fails.

jrajahalme replied:
Changed, thanks!

bimmlerd commented on pkg/container/versioned/value_test.go:
```diff
@@ -0,0 +1,388 @@
+	assert.Len(t, s, 0)
+}
+
+func TestVersionedChaos(t *testing.T) {
```
The race detector is pretty unhappy with this test - `go test -c -race ./pkg/container/versioned && ./versioned.test` gives me immediate hits of the form:

```
==================
WARNING: DATA RACE
Read at 0x00c000074488 by goroutine 148:
  github.com/cilium/cilium/pkg/container/versioned.(*Coordinator).GetVersionHandle()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value.go:342 +0x84
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos.func2()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:337 +0x2c

Previous write at 0x00c000074488 by goroutine 114:
  github.com/cilium/cilium/pkg/container/versioned.(*Coordinator).commit()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value.go:242 +0xac
  github.com/cilium/cilium/pkg/container/versioned.(*Tx).Commit()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value.go:211 +0x3d0
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos.func3()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:367 +0x3ac

Goroutine 148 (running) created at:
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:336 +0x730
  testing.tRunner()
      /usr/lib/go-1.22/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /usr/lib/go-1.22/src/testing/testing.go:1742 +0x40

Goroutine 114 (finished) created at:
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:344 +0x384
  testing.tRunner()
      /usr/lib/go-1.22/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /usr/lib/go-1.22/src/testing/testing.go:1742 +0x40
==================
```

jrajahalme replied:
A `defer` was missing do a lock was released right after closing, rather than at the end of `commit()`, thanks for noticing!

bimmlerd reviewed (commented):
Couple of nits, and a race detector hit which you might want to look at.

mhofstetter reviewed (approved):
`proxy` & `envoy` :heavy_check_mark: 

Even though i haven't looked into the policy / versionHandle changes in detail. The modified functions in `pkg/envoy` are all policy translation logic related (Cilium NP -> Cilium Proxy NPDS).

(IMO this logic would best be placed in a file other than `xds_server.go` - but this is something for  another PR.)

gandro reviewed (approved):
I can't judge the mapstate changes, since I'm not familiar with that area of the code. I'm relying on other folks in sig-policy to review those. 

But the changes to selector cache and the new versioned package seem reasonable to me.

tklauser commented on pkg/endpoint/policy.go:
```diff
@@ -53,7 +53,6 @@ func (e *Endpoint) HasBPFPolicyMap() bool {
 }
 
 // GetNamedPort returns the port for the given name.
-// Must be called with e.mutex NOT held
```
Out of curiosity: why is this no longer the case? Is that an effect of this change or just a leftover from a previous change?

jrajahalme replied:
This is a leftover from a previous change. The comment was necessary earlier when the implementation was taking a lock and/or the called functions depended on this one not holding the endpoint lock. This function does not take the endpoint lock, and the called functions explicitly state that the endpoint lock may be taken while them being called, e.g.:
```
// GetNamedPorts returns a copy of the named ports map. May return nil.
func (ipc *IPCache) GetNamedPorts() (npm types.NamedPortMultiMap) {
	// We must not acquire the IPCache mutex here, as that would establish a lock ordering of
	// Endpoint > IPCache (as endpoint.mutex can be held while calling GetNamedPorts)
	// Since InjectLabels requires IPCache > Endpoint, a deadlock can occur otherwise.
```

tklauser commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,510 @@
+		return ErrStaleVersion
+	}
+	return err
+}
```
nit: let's try to avoid named returns if the type already sufficiently describes the return value, see https://go.dev/wiki/CodeReviewComments#named-result-parameters. `err` is never set anyway:

```suggestion
func (h *VersionHandle) Close() error {
	if h == nil || h.version == invalidVersion {
		return ErrInvalidVersion
	}
	// handle with 'maxVersion' is never closed
	if h.version != maxVersion {
		// Using CompareAndSwap makes sure each handle is closed at most once
		coordinator := h.coordinator.Load()
		if coordinator != nil && h.coordinator.CompareAndSwap(coordinator, nil) {
			runtime.SetFinalizer(h, nil)
			return coordinator.releaseVersion(h.version)
		}
		return ErrStaleVersion
	}
	return nil
}
```

jrajahalme replied:
Done!

tklauser commented on pkg/policy/selectorcache_test.go:
```diff
@@ -144,12 +148,15 @@ func (csu *cachedSelectionUser) IdentitySelectionUpdated(selector CachedSelector
 // Mock CachedSelector for unit testing.
-
+//
+// testCachedSelector is used in isolation so there is no point to impelement versioning for it.
```
nit: typo

```suggestion
// testCachedSelector is used in isolation so there is no point to implement versioning for it.
```

jrajahalme replied:
Fixed, thanks!

tklauser commented on pkg/container/versioned/value.go:
```diff
@@ -0,0 +1,510 @@
+		}
+	}
+	return empty
+}
```
The return value named `empty` will look confusing in godoc because it's not always empty. Let's get rid of the name:

```suggestion
func (v *Value[T]) At(handle *VersionHandle) T {
	if handle != nil {
		version := handle.version
		for node := v.head.Load(); node != nil; node = node.next.Load() {
			if node.versions.contains(version) {
				return node.value
			}
		}
	}
	var empty T
	return empty
}
```

jrajahalme replied:
Done, thanks!

tklauser reviewed (approved):
Some nits and clarifying questions inline, nothing blocking. LGTM for my codeowners.

jrajahalme committed:
policy: Make SelectorCache selections updates transactional

Use the new versioned.Value to make SelectorCache selections updates
transactional. This allows changes to selections of different selectors
to appear at once on a new version published by the selector cache after
all the changes have been completed.

This change allows SelectorCache users to rely on all the selectors to be
in sync with each other, so that if a given identity is selected by two
selectors, that identity appears in the result of the GetSelections(handle)
call for both of the selectors. Previously it was possible that at the
time of the respective GetSelections() call, either of them would have
failed to include the new identity, due to the update of that selector
still being in progress.

With this change the visibility of new identities in the selections of
any cached selector is delayed until all affected selectors have been
updated.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -473,11 +518,17 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted identity.IdentityMap, w
 		}
 	}
+
+	if updated {
```
Do we need to do anything to `txn` if `updated == false`? Rollback or such?

jrajahalme replied:
No, as when there are no changes, there is no state to rollback and the txn can be left un-committed.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1311,21 +1330,14 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 func (e *Endpoint) syncPolicyMap() error {
-	// Apply pending policy map changes first so that desired map is up-to-date before
-	// we diff the maps below.
-	err := e.applyPolicyMapChanges()
```
So, we don't need this anymore because we will apply the pending incremental updates in a subsequent incremental update, right? Thus fixing the leak where incremental updates that come in during regeneration are lost in Envoy?

jrajahalme replied:
This call is moved higher up in the call stack, so that we have consistent view on the mapstate for both bpf map sync and Envoy network policy update, which by design needs to happen before the bpf map sync.
```
--- a/pkg/endpoint/bpf.go
+++ b/pkg/endpoint/bpf.go
@@ -763,6 +763,13 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
 	// Mark the desired policy as ready when done before the lock is released
 	defer e.desiredPolicy.Ready()
 
+	// Apply pending policy map changes so that desired map is up-to-date before
+	// we update network policies and sync the maps below.
+	err = e.applyPolicyMapChanges()
+	if err != nil {
+		return err
+	}
+
 	// We cannot obtain the rules while e.mutex is held, because obtaining
 	// fresh DNSRules requires the IPCache lock (which must not be taken while
 	// holding e.mutex to avoid deadlocks). Therefore, rules are obtained
```

squeed replied:
Awesome, really nice to get this cleaned up.

squeed reviewed (approved):
One minor question, but looks good!

jrajahalme committed:
versioned.Value: Add new container for versioned values

Versioned values can be used for transactional updates where updates in
multiple values become available at the same version number.

Multiple versioned.Values are managed by a common
versioned.Coordinator. Each versioned.Value keeps a linked list of nodes
with the validity range for the value stored on that node. Nodes are
traversed until a value that is valid for the version at the beginning of
the transaction is found, or an empty value is returned otherwise.

Old values are cleaned off when a read transaction is finished by calling
the VersionHandle.Close() function.

In steady-state only the head pointer needs to be dereferenced, which is
the same cost as dereferencing a normal value stored via a pointer. With
churn more nodes may need to be traversed, but that should be
transitional, and on read heavy use cases the average number of traversed
nodes should be close to 1, so the overhead of transactionality should be
minimal.

The list of values is accessed in lock-free manner, so no locks are
needed when getting a value for a version of a valid handle. The entity
updating the values must only allow one goroutine to update any value at
any given time.

This is inspired by earlier work at
https://github.com/openvswitch/ovs/commit/2b7b1427c177d28ec63632646ce896eec7f162b6
and as simplified at
https://github.com/openvswitch/ovs/commit/18721c4a4d49416a0c63ca985988b000fc48f217
and
https://github.com/openvswitch/ovs/commit/44e0c35d981d416d56021e6409dde300148be4cf

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
